Sources/AsyncHTTPClient/HTTPClient.swift

```diff
@@ -423,6 +425,41 @@ extension HTTPClient.Configuration {
             self.read = read
         }
     }
+
+    /// Specifies redirect limit.
+    public struct RedirectLimit {
+        enum Next {
+            case none
+            case loop(visited: Set<URL>)
+            case count(left: Int)
+        }
+
+        var next: Next
+
+        /// No redirect limit.
+        public static let none = RedirectLimit(next: .none)
+        /// Request execution will be stopped if redirect loop is detected.
+        public static let detectLoop = RedirectLimit(next: .loop(visited: Set()))
+        /// Specifies maximum number of redirects for a single request.
+        public static func count(_ value: Int) -> RedirectLimit { return RedirectLimit(next: .count(left: value)) }
+    }
+
+    /// Specifies redirect processing settings.
+    public enum FollowRedirects {
```

Comment:
This public API seems slightly confused, to me... I would approach it as:

```swift
public enum RedirectionPolicy {
    /// Redirects are not allowed/followed.
    case disallow

    /// Redirects are allowed.
    /// limit: The maximum number of redirects allowed.
    /// detectCycles: If true, detect cycles and cancel redirection if detected.
    case allow(limit: Limit, detectCycles: Bool)

    public enum Limit {
        /// Unlimited redirection
        case unlimited

        /// A limited number redirects are allowed
        case limited(count: Int)
    }
}
```

Then, the `Set<URL>` and `count` state would be maintained elsewhere, e.g. a separate `RedirectionState` struct on the request.

In this way we can allow people to set a maximum number of redirects, and also detect cyles. I don't think that is possible in the current design?

Reply:
I think we should also to the docs that the `Set<URL>` case can balloon arbitrary amounts of data into memory so can be used as a DOS vector.

Reply:
feels like we should always set some limit here, both on count and on recursion

Reply:
similar approach to API as one from @ianpartridge

```
public enum FollowRedirects {
    /// Redirects are not followed.
    case no

    /// Redirects are followed.
    /// max: The number of redirects allowed.
    /// allowCycles: If false, detect cycles and cancel redirection if detected.
    case yes(max: Int = 50, allowCycles: Bool = false)
}
```

Reply:
My slight concern with naming the type `FollowRedirects` is that it reads like a boolean and I tend to prefer "...Strategy" or "...Policy" for this kind of enum. No big deal though.

Reply:
@ianpartridge not a strong opinion on my side

Reply:
Great ideas, thanks! I decided to try to incorporate both suggestions - renamed FollowRedirects to RedirectPolicy and `.allow` now has both limits. wdyt?

Reply:
Also, to prevent possible DOS, cycle detection is cut off by max allowed redirects (either cycle will be found or max allowed will fail)

Reply:
looks better imo. i still feel the redirect state really belongs with the request object (as private extension) rather than as peer to it

Reply:
I will do it next, yes

Reply:
Moved redirect state to request